### PR TITLE
Update to NVHPC-23.7 for testing

### PR DESCRIPTION
**Pull request by crtrott**
I went through all the places where we ifdef for NVHPC and checked whether its needed with 23.7. In the tests I unconditionally removed places where the special casing is not needed, in the source code I did a version based check. 

All in all we are going down from over 30 places with issues to 11 FIXME places for NVHPC - 4 of which are a regression in 23.x compared to 22.9.

### CODE REVIEW

File: core/unit_test/TestBitManipulationBuiltins.hpp [hunk 1]
```diff
@@ -828,7 +828,7 @@ struct TestBitCastFunction {
     }
 
 #if defined(KOKKOS_ENABLE_CUDA) && \
-    defined(KOKKOS_COMPILER_NVHPC)  // FIXME_NVHPC
+    defined(KOKKOS_COMPILER_NVHPC)  // FIXME_NVHPC 23.7
```
Comment (dalg24) [inline core/unit_test/TestBitManipulationBuiltins.hpp:831]: Why did you add 23.7 here?

Comment (crtrott) [reply core/unit_test/TestBitManipulationBuiltins.hpp:831]: Book keeping to clarify that yes during this review I checked that compiler. Next time we go through all of this we can update.

Comment (crtrott) [reply core/unit_test/TestBitManipulationBuiltins.hpp:831]: To make that clear: I explicitly tried taking that ifdef out and checked that it still fails.

File: core/unit_test/TestNumericTraits.hpp [hunk 2]
```diff
@@ -396,12 +387,8 @@ TEST(TEST_CATEGORY, numeric_traits_quiet_and_signaling_nan) {
   TestNumericTraits<TEST_EXECSPACE, float, SignalingNaN>();
   TestNumericTraits<TEST_EXECSPACE, double, QuietNaN>();
   TestNumericTraits<TEST_EXECSPACE, double, SignalingNaN>();
-#ifndef KOKKOS_COMPILER_NVHPC  // FIXME_NVHPC:
-  // Unsupported unknown data type 38.
-  // Unsupported unknown data type 38.
-  // Unsupported unknown data type 38.
-  // nvc++-Fatal-/home/projects/x86-64/nvidia/hpc_sdk/Linux_x86_64/22.3/compilers/bin/tools/cpp2
-  // TERMINATED by signal 11
+  // FIXME_NVHPC long double not supported
```
Comment (dalg24) [inline core/unit_test/TestNumericTraits.hpp:390]: Do we consider it a FIXME?

File: algorithms/src/sorting/Kokkos_BinSortPublicAPI.hpp [hunk 3]
```diff
@@ -60,12 +60,8 @@ class BinSort {
         Kokkos::is_view<SrcViewType>::value,
         Kokkos::View<typename SrcViewType::const_data_type,
                      typename SrcViewType::array_layout,
-                     typename SrcViewType::device_type
-#if not defined(KOKKOS_COMPILER_NVHPC)  // FIXME_NVHPC
-                     ,
-                     Kokkos::MemoryTraits<Kokkos::RandomAccess>
-#endif
```
Comment (masterleinad) [inline algorithms/src/sorting/Kokkos_BinSortPublicAPI.hpp:67]: Why do we not check for the version here but in other places?

File: core/unit_test/TestLocalDeepCopy.hpp [hunk 4]
```diff
@@ -906,6 +906,10 @@ void impl_test_local_deepcopy_rangepolicy_rank_7(const int N) {
 
 #if defined(KOKKOS_ENABLE_CXX11_DISPATCH_LAMBDA)
 TEST(TEST_CATEGORY, local_deepcopy_teampolicy_layoutleft) {
+#ifdef KOKKOS_COMPILER_NVHPC  // FIXME_NVHPC 23.7
+  GTEST_SKIP() << "FIXME_NVHPC (?)";
```
Comment (dalg24) [inline core/unit_test/TestLocalDeepCopy.hpp:910]: Fails both on device and host side?

File: core/src/impl/Kokkos_HostThreadTeam.hpp [hunk 5]
```diff
@@ -155,11 +155,7 @@ class HostThreadTeamData {
 
   //----------------------------------------
 
-#ifndef KOKKOS_COMPILER_NVHPC  // FIXME_NVHPC bug in NVHPC regarding constexpr
-                               // constructors used in device code
-  constexpr
-#endif
```
Comment (masterleinad) [inline core/src/impl/Kokkos_HostThreadTeam.hpp:161]: No version check?

**Review by dalg24 — COMMENTED** [review]
Not a big fan of the addition of the "23.7" comments.
Any thought on whether we should do something at configuration time (warn or whatnot) when the version is insufficient?
Also please comment on the strategy, some places you dropped the macro guards and others you check for the compiler version.

File: core/unit_test/TestBitManipulationBuiltins.hpp (hunk 1 above)
Comment (dalg24) [reply core/unit_test/TestBitManipulationBuiltins.hpp:831]: Fine

File: core/unit_test/TestNumericTraits.hpp (hunk 2 above)
Comment (crtrott) [reply core/unit_test/TestNumericTraits.hpp:390]: no, I think its just not supported

File: algorithms/src/sorting/Kokkos_BinSortPublicAPI.hpp (hunk 3 above)
Comment (crtrott) [reply algorithms/src/sorting/Kokkos_BinSortPublicAPI.hpp:67]: Yeah gonna fix it sorry.

File: core/unit_test/TestNumericTraits.hpp (hunk 2 above)
Comment (crtrott) [reply core/unit_test/TestNumericTraits.hpp:390]: Ah sorry yeah should remove the FIXME

File: .jenkins [hunk 6]
```diff
@@ -53,12 +53,12 @@ pipeline {
                               make -j8 && ctest --verbose'''
                     }
                 }
-                stage('CUDA-11.7-NVHPC') {
+                stage('CUDA-12.2-NVHPC') {
```
Comment (masterleinad) [inline .jenkins:56]: Don't forget to update the table with the compiler versions we are testing.

**Review by masterleinad — APPROVED** [review]

File: core/unit_test/TestNumericTraits.hpp [hunk 7]
```diff
@@ -157,7 +157,7 @@ struct TestNumericTraits {
     use_on_device();
   }
   KOKKOS_FUNCTION void operator()(SignalingNaN, int, int& e) const {
-#ifndef KOKKOS_COMPILER_NVHPC  // FIXME_NVHPC
+#ifndef KOKKOS_COMPILER_NVHPC  // FIXME_NVHPC  23.7 nan
```
Comment (dalg24) [inline core/unit_test/TestNumericTraits.hpp:160]:
```suggestion
#ifndef KOKKOS_COMPILER_NVHPC  // FIXME_NVHPC 23.7 nan
```

File: core/unit_test/TestLocalDeepCopy.hpp [hunk 8]
```diff
@@ -907,7 +907,12 @@ void impl_test_local_deepcopy_rangepolicy_rank_7(const int N) {
 #if defined(KOKKOS_ENABLE_CXX11_DISPATCH_LAMBDA)
 TEST(TEST_CATEGORY, local_deepcopy_teampolicy_layoutleft) {
   using ExecSpace = TEST_EXECSPACE;
-  using ViewType  = Kokkos::View<double********, Kokkos::LayoutLeft, ExecSpace>;
+#if defined(KOKKOS_ENABLE_CUDA) && \
+    defined(KOKKOS_COMPILER_NVHPC)  // FIXME_NVHPC 23.7
+  if (std::is_same_v<ExecSpace, Kokkos::Cuda>)
+    GTEST_SKIP() << "FIXME_NVHPC (?)";
```
Comment (dalg24) [inline core/unit_test/TestLocalDeepCopy.hpp:913]: The custom skip message is not great.

**Review by dalg24 — APPROVED** [review]

File: core/unit_test/TestLocalDeepCopy.hpp (hunk 8 above)
Comment (crtrott) [reply core/unit_test/TestLocalDeepCopy.hpp:913]: Copied from elsewhere but I can update. Need to anyway fix OpenACC build

**Comment by crtrott** [comment]
@seyonglee I am updating testing here also for OpenACC to NVHPC 23.7, otherwise I would need to ifdef tests for OpenACC. Any objections?

**Comment by seyonglee** [comment]
> @seyonglee I am updating testing here also for OpenACC to NVHPC 23.7, otherwise I would need to ifdef tests for OpenACC. Any objections?

I don't have access to NVHPC 23.7 and thus cannot check whether NVHPC 23.7 will cause any troubles for OpenACC or not. However, if you didn't find any issues, let's upgrade the OpenACC testing configuration to NVHPC 23.7 too.

**Comment by crtrott** [comment]
Retest this please.

File: .jenkins [hunk 9]
```diff
@@ -27,17 +27,18 @@ pipeline {
         }
         stage('Build') {
             parallel {
-                stage('OPENACC-NVHPC-CUDA-11.6') {
+                stage('OPENACC-NVHPC-CUDA-12.2') {
                     agent {
                         dockerfile {
                             filename 'Dockerfile.nvhpc'
                             dir 'scripts/docker'
+                            additionalBuildArgs '--build-arg BASE=nvcr.io/nvidia/nvhpc:23.7-devel-cuda12.2-ubuntu20.04'
```
Comment (dalg24) [inline .jenkins:35]: Please update https://github.com/kokkos/kokkos/blob/1a3ea28f6e97b4c9dd2c8ceed53ad58ed5f94dfe/scripts/docker/Dockerfile.nvhpc#L1

Comment (dalg24) [reply .jenkins:35]: It is fine to specify the build argument here but the rule of thumb is that images built without build arguments should also be usable.

File: core/unit_test/TestLocalDeepCopy.hpp [hunk 10]
```diff
@@ -907,7 +907,13 @@ void impl_test_local_deepcopy_rangepolicy_rank_7(const int N) {
 #if defined(KOKKOS_ENABLE_CXX11_DISPATCH_LAMBDA)
 TEST(TEST_CATEGORY, local_deepcopy_teampolicy_layoutleft) {
   using ExecSpace = TEST_EXECSPACE;
-  using ViewType  = Kokkos::View<double********, Kokkos::LayoutLeft, ExecSpace>;
+#if defined(KOKKOS_ENABLE_CUDA) && \
+    defined(KOKKOS_COMPILER_NVHPC)  // FIXME_NVHPC 23.7
+  if (std::is_same_v<ExecSpace, Kokkos::Cuda>)
+    GTEST_SKIP()
+        << "FIXME_NVHPC : Compiler bug affecting subviews of high rank Views";
```
Comment (dalg24) [inline core/unit_test/TestLocalDeepCopy.hpp:914]: The `FIXME_NVHPC :` prefix is unnecessary but the message is much better now.

**Comment by crtrott** [comment]
Gonna rebase on top of #6385 to fix the CI errors.

**Comment by crtrott** [comment]
Damien: "merge, what could possibly go wrong"
